packages/twenty-server/src/modules/contact-creation-manager/services/create-company.service.ts

```diff
@@ -1,108 +1,189 @@
+import uniqBy from 'lodash.uniqby';
 
-import { InjectObjectMetadataRepository } from 'src/engine/object-metadata-repository/object-metadata-repository.decorator';
-import { CompanyRepository } from 'src/modules/company/repositories/company.repository';
 import { CompanyWorkspaceEntity } from 'src/modules/company/standard-objects/company.workspace-entity';
 import { extractDomainFromLink } from 'src/modules/contact-creation-manager/utils/extract-domain-from-link.util';
 import { getCompanyNameFromDomainName } from 'src/modules/contact-creation-manager/utils/get-company-name-from-domain-name.util';
+import { FieldActorSource } from 'src/engine/metadata-modules/field-metadata/composite-types/actor.composite-type';
+import { WorkspaceMemberWorkspaceEntity } from 'src/modules/workspace-member/standard-objects/workspace-member.workspace-entity';
+import { TwentyORMGlobalManager } from 'src/engine/twenty-orm/twenty-orm-global.manager';
+import { computeDisplayName } from 'src/utils/compute-display-name';
+import { WorkspaceRepository } from 'src/engine/twenty-orm/repository/workspace.repository';
+
+type CompanyToCreate = {
+  domainName: string;
+  createdBySource: FieldActorSource;
+  createdByWorkspaceMember?: WorkspaceMemberWorkspaceEntity | null;
+};
+
 @Injectable()
 export class CreateCompanyService {
   private readonly httpService: AxiosInstance;
 
-  constructor(
-    @InjectObjectMetadataRepository(CompanyWorkspaceEntity)
-    private readonly companyRepository: CompanyRepository,
-  ) {
+  constructor(private readonly twentyORMGlobalManager: TwentyORMGlobalManager) {
     this.httpService = axios.create({
       baseURL: 'https://companies.twenty.com',
     });
   }
 
   async createCompanies(
-    domainNames: string[],
+    companies: CompanyToCreate[],
     workspaceId: string,
-    companyDomainNameColumnName: string,
     transactionManager?: EntityManager,
   ): Promise<{
     [domainName: string]: string;
   }> {
-    if (domainNames.length === 0) {
+    if (companies.length === 0) {
       return {};
     }
 
-    const uniqueDomainNames = [...new Set(domainNames)];
-
-    const existingCompanies =
-      await this.companyRepository.getExistingCompaniesByDomainNames(
-        uniqueDomainNames,
+    const companyRepository =
+      await this.twentyORMGlobalManager.getRepositoryForWorkspace(
         workspaceId,
-        companyDomainNameColumnName,
-        transactionManager,
+        CompanyWorkspaceEntity,
       );
 
-    const companiesObject = existingCompanies.reduce(
-      (
-        acc: {
-          [domainName: string]: string;
-        },
-        company: {
-          domainName: string;
-          id: string;
-        },
-      ) => ({
-        ...acc,
-        [extractDomainFromLink(company.domainName)]: company.id,
-      }),
-      {},
+    // Avoid creating duplicate companies
+    const uniqueCompanies = uniqBy(companies, 'domainName');
+    const conditions = uniqueCompanies.map((companyToCreate) => ({
```

Comment:
To check performances 🤔

Reply:
It's actually run in a job, but I agree, maybe we can optimize this, I've try to improve it a bit already